File: tests/fixtures/realworld/kotlin/complex/src/main/kotlin/com/okfgen/service/repo/UserRepository.kt
package com.okfgen.service.repo

import com.okfgen.service.model.Paginated
import com.okfgen.service.model.User

/**
 * Generic repository interface for CRUD operations.
 */
interface Repository<T : Comparable<T>> {
    fun save(entity: T): T
    fun findById(id: String): T?
    fun findAll(): List<T>
    fun deleteById(id: String): Boolean
    fun count(): Int
}

/**
 * In-memory implementation of Repository for User entities.
 */
class UserRepository : Repository<User> {
    private val store = mutableMapOf<String, User>()

    override fun save(user: User): User {
        store[user.id] = user
        return user
    }

    override fun findById(id: String): User? = store[id]

    override fun findAll(): List<User> = store.values.toList()

    override fun deleteById(id: String): Boolean = store.remove(id) != null

    override fun count(): Int = store.size

    /**
     * Returns users sorted by creation date with pagination.
     */
    fun listPaginated(page: Int, pageSize: Int): Paginated<User> {
        val all = store.values.sorted()
        val total = all.size
        val start = maxOf(0, (page - 1) * pageSize)
        val items = all.drop(start).take(pageSize)
        return Paginated(items = items, total = total, page = page, pageSize = pageSize)
    }

    /**
     * Finds users by a predicate.
     */
    inline fun findBy(predicate: (User) -> Boolean): List<User> =
        store.values.filter(predicate)
}

File: tests/fixtures/realworld/swift/complex/Sources/Service/Repositories/UserRepository.swift
import Foundation

/// Protocol defining CRUD operations for a generic entity type.
public protocol Repository<T> {
    associatedtype T: Codable & Comparable
    func save(_ entity: T) -> T
    func findById(_ id: String) -> T?
    func findAll() -> [T]
    func deleteById(_ id: String) -> Bool
    func count() -> Int
}

/// Possible states of a user account lifecycle.
public enum AccountStatus: String, Codable {
    case active
    case suspended
    case archived
}

/// In-memory implementation of Repository for User entities.
public class UserRepository: Repository {
    public typealias T = User
    private var store: [String: User] = [:]

    public init() {}

    public func save(_ user: User) -> User {
        store[user.id] = user
        return user
    }

    public func findById(_ id: String) -> User? {
        store[id]
    }

    public func findAll() -> [User] {
        Array(store.values)
    }

    public func deleteById(_ id: String) -> Bool {
        store.removeValue(forKey: id) != nil
    }

    public func count() -> Int {
        store.count
    }

    /// Returns users sorted by creation date with pagination.
    public func listPaginated(page: Int, pageSize: Int) -> Paginated<User> {
        let all = store.values.sorted()
        let total = all.count
        let start = max(0, (page - 1) * pageSize)
        let items = Array(all.dropFirst(start).prefix(pageSize))
        return Paginated(items: items, total: total, page: page, pageSize: pageSize)
    }
}
